File: src/android/MapboxPlugin.java
package com.outsystems.plugins.mapbox;

import android.content.Context;
import android.content.res.Resources;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.mapbox.geojson.Point;
import com.mapbox.maps.CameraOptions;
import com.mapbox.maps.CameraState;
import com.mapbox.maps.MapView;
import com.mapbox.maps.MapboxMap;
import com.mapbox.maps.ScreenCoordinate;
import com.mapbox.maps.Style;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONException;
import org.json.JSONObject;

public class MapboxPlugin extends CordovaPlugin {
  private static final String TOKEN_RESOURCE = "mapbox_access_token";

  private MapView mapView;
  private MapboxMap mapboxMap;
  private CallbackContext regionDidChangeCallback;

  @Override
  public boolean execute(String action, CordovaArgs args, CallbackContext callbackContext) throws JSONException {
    if ("show".equals(action)) {
      show(args.getJSONObject(0), callbackContext);
      return true;
    }
    if ("hide".equals(action)) {
      hide(callbackContext);
      return true;
    }
    if ("setCenter".equals(action)) {
      setCenter(args.getJSONObject(0), callbackContext);
      return true;
    }
    if ("getCenter".equals(action)) {
      getCenter(callbackContext);
      return true;
    }
    if ("setZoomLevel".equals(action)) {
      setZoomLevel(args.getJSONObject(0), callbackContext);
      return true;
    }
    if ("getZoomLevel".equals(action)) {
      getZoomLevel(callbackContext);
      return true;
    }
    if ("setTilt".equals(action)) {
      setTilt(args.getJSONObject(0), callbackContext);
      return true;
    }
    if ("getTilt".equals(action)) {
      getTilt(callbackContext);
      return true;
    }
    if ("animateCamera".equals(action)) {
      animateCamera(args.getJSONObject(0), callbackContext);
      return true;
    }
    if ("getBounds".equals(action)) {
      getBounds(callbackContext);
      return true;
    }
    if ("convertCoordinate".equals(action)) {
      convertCoordinate(args.getJSONObject(0), callbackContext);
      return true;
    }
    if ("convertPoint".equals(action)) {
      convertPoint(args.getJSONObject(0), callbackContext);
      return true;
    }
    if ("onRegionDidChange".equals(action)) {
      this.regionDidChangeCallback = callbackContext;
      PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
      result.setKeepCallback(true);
      callbackContext.sendPluginResult(result);
      return true;
    }
    if ("onRegionWillChange".equals(action) || "onRegionIsChanging".equals(action)
        || "addMarkers".equals(action) || "removeAllMarkers".equals(action)
        || "addMarkerCallback".equals(action) || "addPolygon".equals(action)
        || "addGeoJSON".equals(action) || "setBounds".equals(action)) {
      callbackContext.error(action + " is not implemented in this O11 starter. Implement with Mapbox v11 Annotation/Style APIs before production use.");
      return true;
    }
    return false;
  }

  private void show(final JSONObject options, final CallbackContext callbackContext) {
    cordova.getActivity().runOnUiThread(new Runnable() {
      @Override
      public void run() {
        try {
          String accessToken = readAccessToken();
          if (accessToken == null || accessToken.length() == 0) {
            callbackContext.error("ACCESS_TOKEN is missing.");
            return;
          }

          Context context = cordova.getActivity();
          com.mapbox.maps.MapboxOptions.INSTANCE.setAccessToken(accessToken);

          JSONObject center = options.optJSONObject("center");
          double lat = center == null ? 0.0 : center.optDouble("lat", 0.0);
          double lng = center == null ? 0.0 : center.optDouble("lng", 0.0);
          double zoom = options.optDouble("zoomLevel", center == null ? 0.0 : 10.0);

          CameraOptions camera = new CameraOptions.Builder()
              .center(Point.fromLngLat(lng, lat))
              .zoom(zoom)
              .build();

          mapView = new MapView(context);
          mapboxMap = mapView.getMapboxMap();
          mapboxMap.setCamera(camera);
          mapboxMap.loadStyle(resolveStyle(options.optString("style", "streets")));

          JSONObject margins = options.optJSONObject("margins");
          int left = margins == null ? 0 : margins.optInt("left", 0);
          int top = margins == null ? 0 : margins.optInt("top", 0);
          int right = margins == null ? 0 : margins.optInt("right", 0);
          int bottom = margins == null ? 0 : margins.optInt("bottom", 0);

          int webViewWidth = webView.getView().getWidth();
          int webViewHeight = webView.getView().getHeight();
          FrameLayout layout = (FrameLayout) webView.getView().getParent();
          FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(webViewWidth - left - right, webViewHeight - top - bottom);
          params.setMargins(left, top, right, bottom);
          mapView.setLayoutParams(params);
          layout.addView(mapView);
          callbackContext.success();
        } catch (Throwable t) {
          callbackContext.error(t.getMessage());
        }
      }
    });
  }

  private void hide(final CallbackContext callbackContext) {
    cordova.getActivity().runOnUiThread(new Runnable() {
      @Override
      public void run() {
        if (mapView != null) {
          ViewGroup parent = (ViewGroup) mapView.getParent();
          if (parent != null) {
            parent.removeView(mapView);
          }
          mapView.onDestroy();
          mapView = null;
          mapboxMap = null;
        }
        callbackContext.success();
      }
    });
  }

  private void setCenter(JSONObject options, CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    Point point = Point.fromLngLat(options.optDouble("lng"), options.optDouble("lat"));
    mapboxMap.setCamera(new CameraOptions.Builder().center(point).build());
    sendRegionDidChange();
    callbackContext.success();
  }

  private void getCenter(CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    Point center = mapboxMap.getCameraState().getCenter();
    JSONObject json = new JSONObject();
    try {
      json.put("lat", center.latitude());
      json.put("lng", center.longitude());
      callbackContext.success(json);
    } catch (JSONException e) {
      callbackContext.error(e.getMessage());
    }
  }

  private void setZoomLevel(JSONObject options, CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    mapboxMap.setCamera(new CameraOptions.Builder().zoom(options.optDouble("level")).build());
    sendRegionDidChange();
    callbackContext.success();
  }

  private void getZoomLevel(CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    callbackContext.success(String.valueOf(mapboxMap.getCameraState().getZoom()));
  }

  private void setTilt(JSONObject options, CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    mapboxMap.setCamera(new CameraOptions.Builder().pitch(options.optDouble("pitch", 0.0)).build());
    sendRegionDidChange();
    callbackContext.success();
  }

  private void getTilt(CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    callbackContext.success(String.valueOf(mapboxMap.getCameraState().getPitch()));
  }

  private void animateCamera(JSONObject options, CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    JSONObject target = options.optJSONObject("target");
    CameraOptions.Builder builder = new CameraOptions.Builder();
    if (target != null) {
      builder.center(Point.fromLngLat(target.optDouble("lng"), target.optDouble("lat")));
    }
    if (options.has("zoomLevel")) builder.zoom(options.optDouble("zoomLevel"));
    if (options.has("tilt")) builder.pitch(options.optDouble("tilt"));
    if (options.has("bearing")) builder.bearing(options.optDouble("bearing"));
    mapboxMap.setCamera(builder.build());
    sendRegionDidChange();
    callbackContext.success();
  }

  private void getBounds(CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    try {
      Point sw = mapboxMap.coordinateForPixel(new ScreenCoordinate(0.0, mapView.getHeight()));
      Point ne = mapboxMap.coordinateForPixel(new ScreenCoordinate(mapView.getWidth(), 0.0));
      JSONObject json = new JSONObject();
      json.put("sw_lat", sw.latitude());
      json.put("sw_lng", sw.longitude());
      json.put("ne_lat", ne.latitude());
      json.put("ne_lng", ne.longitude());
      callbackContext.success(json);
    } catch (JSONException e) {
      callbackContext.error(e.getMessage());
    }
  }

  private void convertCoordinate(JSONObject options, CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    try {
      ScreenCoordinate point = mapboxMap.pixelForCoordinate(Point.fromLngLat(options.optDouble("lng"), options.optDouble("lat")));
      JSONObject json = new JSONObject();
      json.put("x", point.getX());
      json.put("y", point.getY());
      callbackContext.success(json);
    } catch (JSONException e) {
      callbackContext.error(e.getMessage());
    }
  }

  private void convertPoint(JSONObject options, CallbackContext callbackContext) {
    if (!ensureMap(callbackContext)) return;
    try {
      Point point = mapboxMap.coordinateForPixel(new ScreenCoordinate(options.optDouble("x"), options.optDouble("y")));
      JSONObject json = new JSONObject();
      json.put("lat", point.latitude());
      json.put("lng", point.longitude());
      callbackContext.success(json);
    } catch (JSONException e) {
      callbackContext.error(e.getMessage());
    }
  }

  private boolean ensureMap(CallbackContext callbackContext) {
    if (mapView == null || mapboxMap == null) {
      callbackContext.error("Mapbox.show must be called first.");
      return false;
    }
    return true;
  }

  private String readAccessToken() {
    try {
      Resources resources = cordova.getActivity().getResources();
      int id = resources.getIdentifier(TOKEN_RESOURCE, "string", cordova.getActivity().getPackageName());
      return id == 0 ? null : resources.getString(id);
    } catch (Resources.NotFoundException e) {
      return null;
    }
  }

  private String resolveStyle(String requested) {
    if ("light".equalsIgnoreCase(requested)) return Style.LIGHT;
    if ("dark".equalsIgnoreCase(requested)) return Style.DARK;
    if ("satellite".equalsIgnoreCase(requested)) return Style.SATELLITE;
    if ("hybrid".equalsIgnoreCase(requested)) return Style.SATELLITE_STREETS;
    if ("streets".equalsIgnoreCase(requested)) return Style.MAPBOX_STREETS;
    if (requested != null && requested.startsWith("mapbox://")) return requested;
    return Style.MAPBOX_STREETS;
  }

  private void sendRegionDidChange() {
    if (regionDidChangeCallback == null || mapboxMap == null) return;
    try {
      CameraState camera = mapboxMap.getCameraState();
      JSONObject json = new JSONObject();
      json.put("lat", camera.getCenter().latitude());
      json.put("lng", camera.getCenter().longitude());
      json.put("camPitch", camera.getPitch());
      json.put("camHeading", camera.getBearing());
      PluginResult result = new PluginResult(PluginResult.Status.OK, json);
      result.setKeepCallback(true);
      regionDidChangeCallback.sendPluginResult(result);
    } catch (JSONException ignored) {
    }
  }

  @Override
  public void onPause(boolean multitasking) {
    if (mapView != null) mapView.onStop();
  }

  @Override
  public void onResume(boolean multitasking) {
    if (mapView != null) mapView.onStart();
  }

  @Override
  public void onDestroy() {
    if (mapView != null) mapView.onDestroy();
  }
}

File: src/android/MapboxPlugin.kt
package com.outsystems.mapbox

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.view.Gravity
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import com.mapbox.geojson.Point
import com.mapbox.common.MapboxOptions
import com.mapbox.maps.CameraOptions
import com.mapbox.maps.MapView
import com.mapbox.maps.Style
import com.mapbox.maps.plugin.annotation.annotations
import com.mapbox.maps.plugin.annotation.generated.PointAnnotation
import com.mapbox.maps.plugin.annotation.generated.PointAnnotationManager
import com.mapbox.maps.plugin.annotation.generated.PointAnnotationOptions
import com.mapbox.maps.plugin.annotation.generated.createPointAnnotationManager
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.json.JSONArray
import org.json.JSONObject

open class MapboxPlugin : CordovaPlugin() {
    private var mapView: MapView? = null
    private var rootView: FrameLayout? = null
    private var annotationManager: PointAnnotationManager? = null
    private val markers = mutableMapOf<String, PointAnnotation>()
    private val defaultMarkerBitmap: Bitmap by lazy { createDefaultMarkerBitmap() }

    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        when (action) {
            "initialize" -> initialize(args.optJSONObject(0) ?: JSONObject(), callbackContext)
            "setViewport" -> setViewport(args.optJSONObject(0) ?: JSONObject(), callbackContext)
            "setCamera" -> setCamera(args.optJSONObject(0) ?: JSONObject(), callbackContext)
            "addMarker" -> addMarker(args.optJSONObject(0) ?: JSONObject(), callbackContext)
            "removeMarker" -> removeMarker(args.optJSONObject(0) ?: JSONObject(), callbackContext)
            "clearMarkers" -> clearMarkers(callbackContext)
            "getCamera" -> getCamera(callbackContext)
            "close" -> close(callbackContext)
            else -> return false
        }
        return true
    }

    private fun initialize(options: JSONObject, callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            try {
                val token = options.optString("token")
                    .ifBlank { preferences.getString("MAPBOX_ACCESS_TOKEN", "") }
                    .ifBlank { androidString("mapbox_access_token") }
                    .takeUnless { it == "__MAPBOX_ACCESS_TOKEN_NOT_SET__" }
                    ?: ""

                if (token.isBlank()) {
                    callback.error("Mapbox access token is required.")
                    return@runOnUiThread
                }

                MapboxOptions.accessToken = token

                val styleUrl = options.optString("styleUrl", Style.MAPBOX_STREETS)
                val latitude = options.optDouble("latitude", 0.0)
                val longitude = options.optDouble("longitude", 0.0)

                if (!isValidLatitude(latitude) || !isValidLongitude(longitude)) {
                    callback.error("Invalid coordinates: latitude must be in [-90, 90], longitude in [-180, 180].")
                    return@runOnUiThread
                }

                val zoom = options.optDouble("zoom", 12.0)

                closeInternal()

                val activity = cordova.activity
                val decor = activity.window.decorView as ViewGroup

                rootView = FrameLayout(activity).apply {
                    setBackgroundColor(Color.WHITE)
                    layoutParams = layoutParamsFromOptions(options)
                }

                mapView = MapView(activity).apply {
                    layoutParams = FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.MATCH_PARENT,
                        FrameLayout.LayoutParams.MATCH_PARENT,
                        Gravity.CENTER
                    )
                }

                rootView!!.addView(mapView)
                if (!options.optBoolean("inline", false)) {
                    rootView!!.addView(Button(activity).apply {
                        text = "Close"
                        setOnClickListener { closeInternal() }
                        layoutParams = FrameLayout.LayoutParams(
                            FrameLayout.LayoutParams.WRAP_CONTENT,
                            FrameLayout.LayoutParams.WRAP_CONTENT,
                            Gravity.TOP or Gravity.END
                        ).apply {
                            topMargin = 48
                            rightMargin = 24
                        }
                    })
                }
                decor.addView(rootView)

                mapView!!.mapboxMap.setCamera(
                    CameraOptions.Builder()
                        .center(Point.fromLngLat(longitude, latitude))
                        .zoom(zoom)
                        .build()
                )

                mapView!!.mapboxMap.loadStyle(styleUrl) {
                    annotationManager = mapView!!.annotations.createPointAnnotationManager()
                    callback.success(JSONObject().put("status", "initialized"))
                }
            } catch (e: Exception) {
                callback.error(e.message ?: "Failed to initialize Mapbox map.")
            }
        }
    }

    private fun setViewport(options: JSONObject, callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            val view = rootView
            if (view == null) {
                callback.error("Map is not initialized.")
                return@runOnUiThread
            }

            view.layoutParams = layoutParamsFromOptions(options)
            view.requestLayout()
            callback.success()
        }
    }

    private fun isValidLatitude(lat: Double) = lat.isFinite() && lat in -90.0..90.0

    private fun isValidLongitude(lon: Double) = lon.isFinite() && lon in -180.0..180.0

    private fun setCamera(options: JSONObject, callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            val map = mapView?.mapboxMap
            if (map == null) {
                callback.error("Map is not initialized.")
                return@runOnUiThread
            }

            val camLat = options.optDouble("latitude")
            val camLng = options.optDouble("longitude")
            if (!isValidLatitude(camLat) || !isValidLongitude(camLng)) {
                callback.error("Invalid coordinates: latitude must be in [-90, 90], longitude in [-180, 180].")
                return@runOnUiThread
            }

            val camera = CameraOptions.Builder()
                .center(Point.fromLngLat(camLng, camLat))
                .zoom(options.optDouble("zoom", map.cameraState.zoom))
                .bearing(options.optDouble("bearing", map.cameraState.bearing))
                .pitch(options.optDouble("pitch", map.cameraState.pitch))
                .build()

            map.setCamera(camera)
            callback.success()
        }
    }

    private fun addMarker(options: JSONObject, callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            val manager = annotationManager
            if (manager == null) {
                callback.error("Map is not initialized.")
                return@runOnUiThread
            }

            val id = options.optString("id").ifBlank { System.currentTimeMillis().toString() }
            markers[id]?.let { manager.delete(it) }

            val markerLat = options.optDouble("latitude")
            val markerLng = options.optDouble("longitude")
            if (!isValidLatitude(markerLat) || !isValidLongitude(markerLng)) {
                callback.error("Invalid coordinates: latitude must be in [-90, 90], longitude in [-180, 180].")
                return@runOnUiThread
            }

            val markerOptions = PointAnnotationOptions()
                .withPoint(Point.fromLngLat(markerLng, markerLat))
                .withIconImage(defaultMarkerBitmap)

            val marker = manager.create(markerOptions)
            markers[id] = marker
            callback.success(JSONObject().put("id", id))
        }
    }

    private fun removeMarker(options: JSONObject, callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            val id = options.optString("id")
            val marker = markers.remove(id)
            val manager = annotationManager

            if (marker != null && manager != null) {
                manager.delete(marker)
            }

            callback.success()
        }
    }

    private fun clearMarkers(callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            annotationManager?.deleteAll()
            markers.clear()
            callback.success()
        }
    }

    private fun getCamera(callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            val map = mapView?.mapboxMap
            if (map == null) {
                callback.error("Map is not initialized.")
                return@runOnUiThread
            }

            val state = map.cameraState
            callback.success(
                JSONObject()
                    .put("latitude", state.center.latitude())
                    .put("longitude", state.center.longitude())
                    .put("zoom", state.zoom)
                    .put("bearing", state.bearing)
                    .put("pitch", state.pitch)
            )
        }
    }

    private fun close(callback: CallbackContext) {
        cordova.activity.runOnUiThread {
            closeInternal()
            callback.success()
        }
    }

    private fun closeInternal() {
        annotationManager = null
        markers.clear()
        mapView?.onStop()
        mapView?.onDestroy()
        rootView?.let { view ->
            (view.parent as? ViewGroup)?.removeView(view)
        }
        mapView = null
        rootView = null
    }

    override fun onResume(multitasking: Boolean) {
        super.onResume(multitasking)
        mapView?.onStart()
    }

    override fun onPause(multitasking: Boolean) {
        mapView?.onStop()
        super.onPause(multitasking)
    }

    override fun onDestroy() {
        closeInternal()
        super.onDestroy()
    }

    private fun createDefaultMarkerBitmap(): Bitmap {
        val width = 72
        val height = 96
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG)

        val pin = Path().apply {
            moveTo(width / 2f, height - 6f)
            cubicTo(10f, 56f, 8f, 16f, width / 2f, 16f)
            cubicTo(width - 8f, 16f, width - 10f, 56f, width / 2f, height - 6f)
            close()
        }

        paint.color = Color.rgb(220, 38, 38)
        canvas.drawPath(pin, paint)
        paint.color = Color.WHITE
        canvas.drawCircle(width / 2f, 40f, 13f, paint)
        return bitmap
    }

    private fun androidString(name: String): String {
        val activity = cordova.activity
        val resourceId = activity.resources.getIdentifier(name, "string", activity.packageName)
        return if (resourceId == 0) "" else activity.getString(resourceId)
    }

    private fun layoutParamsFromOptions(options: JSONObject): FrameLayout.LayoutParams {
        if (!options.optBoolean("inline", false)) {
            return FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
        }

        val density = cordova.activity.resources.displayMetrics.density
        val width = (options.optDouble("width") * density).toInt().coerceAtLeast(1)
        val height = (options.optDouble("height") * density).toInt().coerceAtLeast(1)
        val x = (options.optDouble("x") * density).toInt()
        val y = (options.optDouble("y") * density).toInt()

        return FrameLayout.LayoutParams(width, height).apply {
            leftMargin = x
            topMargin = y
        }
    }
}
